test/jdk/java/net/URLConnection/RequestProperties.java

```diff
@@ -79,4 +84,77 @@ private static boolean hasFtp() {
+            expectIllegalStateException(
+                    () -> {
+                        conn.setRequestProperty("foo", "bar");
+                        return null;
+                    }, "setRequestProperty on " + conn.getClass().getName()
+                            + " for " + url + " was expected to throw"
+                            + " IllegalStateException, but didn't");
+            // test addRequestProperty
+            expectIllegalStateException(
+                    () -> {
+                        conn.addRequestProperty("foo", "bar");
+                        return null;
+                    }, "addRequestProperty on " + conn.getClass().getName()
+                            + " for " + url + " was expected to throw"
+                            + " IllegalStateException, but didn't");
+            // test getRequestProperty
+            expectIllegalStateException(
+                    () -> {
+                        conn.getRequestProperty("foo");
+                        return null;
+                    }, "getRequestProperty on " + conn.getClass().getName()
+                            + " for " + url + " was expected to throw"
+                            + " IllegalStateException, but didn't");
+            // test getRequestProperties
+            expectIllegalStateException(
+                    () -> {
+                        conn.getRequestProperties();
+                        return null;
+                    }, "getRequestProperties on " + conn.getClass().getName()
+                            + " for " + url + " was expected to throw"
+                            + " IllegalStateException, but didn't");
+        } finally {
+            try {
+                conn.getInputStream().close();
+            } catch (Exception e) {
+                // ignore
+            }
+        }
+
+    }
+
+    /**
+     * Calls the {@code operation} and expects it to throw an {@link IllegalStateException}.
+     * If no such exception is thrown then this method throws a {@link RuntimeException}
+     * with the passed {@code unmetExpectationErrorMessage} as the exception's message.
+     *
+     * @param operation                    The operation to invoke
+     * @param unmetExpectationErrorMessage The error message to be set in the
+     *                                     RuntimeException that will be thrown if the operation
+     *                                     doesn't result in an IllegalStateException
+     */
+    private static void expectIllegalStateException(final Callable<Void> operation,
+                                                    final String unmetExpectationErrorMessage)
+            throws Exception {
+        try {
+            operation.call();
+            // the expected IllegalStateException wasn't throw
+            throw new RuntimeException(unmetExpectationErrorMessage);
+        } catch (IllegalStateException ise) {
+            // expected
+        }
+    }
```

Comment:
Given that the test update is significant then it makes me wonder if we should just move to TestNG while we're at it. It would avoid needing to add expectIllegalStateException.
Minor inconsistency is that is the existing test now has "NPE" as the suffix, the new test "IllegalStateException" as the suffix. Also can we find a better name for "unmetExpectationErrorMessage" as it's not clear to maintainers what "unmet" means.

Reply:
> Given that the test update is significant then it makes me wonder if we should just move to TestNG while we're at it. It would avoid needing to add expectIllegalStateException.

I wasn't sure if it was OK to do it as part of this change. If that's fine, then I will go ahead and change this whole test class into TestNG based test.